### ISSUE TITLE
Support use-case-driven parsing through stage1-only tape

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#923]
Right now, simdjson parses JSON to a generic, fully-parsed, fully-validated JSON document. Because don't know what types to expect, and what format it needs to be output, stage 2 takes performance hits, does unnecessary work, and the user still has to do *more* work to convert our format to theirs.

I propose the parser produce the list of structural indexes (call stage 1) without parsing or validating anything numbers, strings and atoms. As they ask us for values with `get<double>()`, etc., we validate and parse into the exact format they want. By doing this, parsing can get inlined, branches removed, and simpler parsing routines used for each value.

The Problem
-------

The user has no way to guide the parser by telling us which values they need, what types they are, and how they want them output. As a result, while the current generic stage 2 is a marvel of performance for what it does, it does unnecessary work, takes performance hits from genericity, and has to produce intermediate values that the user has to do extra work to get into their format.

1. **Intermediate Format:** Users often want to take our data and store it in some known format (passing it to another language, storing in a struct or array, etc.). It would be nice if they could do so directly rather than mediate that through a tape format. Even the production of a 64-bit int when the user wants a 16-bit value is unnecessary.
2. **Unused Values:** A lot of the work we do producing values is wasted, because the user often doesn't use everything in the document.
3. **Variable-Sized Tape:** The current document format has variable element size (2 for numbers, 1 for everything else). This makes the location of the next element less predictable, making navigation, counts, and other things difficult.
4. **Known Types:** Users almost always know exactly what type a given value must be--they express it with `get<>()`. Since our parsing is divorced from the use, we have to unnecessarily branch-and-parse everything, no matter what.
5. **Known Subtypes:** Often numbers and strings are restricted in ways that can speed parsing. The int64_t vs. uint64_t vs. double distinction, for example, can cut out a lot of branching. More than that, knowing it's a uint8_t, or even that it will always be 0-9 (one digit), can drastically speed up parsing of a given number. Even strings can be done easily: sometimes users want
6. **Raw Input:** Sometimes users want to know the original string that produced a value, for things like round tripping or presenting an error message to the user that reflects what they actually typed. Right now, we lose information when we parse floating-point numbers and escaped strings, so the user can't get the original string back even if they want to.
7. **Known Array Types:** Arrays are almost universally all the same type, yet we don't (and can't) vectorize their parsing. This could admittedly be done reasonably well for arrays of numbers and strings if we put our minds to it, but there's no way to for us to bake a routine to parse an array of objects that all have exactly the same fields.

Proposal
-------

I propose a `simdjson::stream::parser.start(padded_string)` interface, which runs stage 1 on a document, and returns an element-like API with access to both the structural marks and the buffer. The difference is, the interface will validate types when you do get<object/array>(), and will both validate and parse when you do get<double>() and get<string>(). This code will be inlined as part of the user's JSON reading code.

### Example: Array of Integers

This will allow things like array struct parsing to be inlined and largely de-branched:

```c++
simdjson::stream::parse parser;
uint64_t total = 0;
uint64_t count = 0;
for (auto value : parser.parse(json)) {
  total += array.get<uint64_t>();
  count++;
}
printf("Average: %lu\n", total / count);
```

get<uint64_t> in this case would look like:

```c++
simdjson_result<uint64_t> stream::element::get<uint64_t>() {
  uint32_t i = *element_structural_index;
  uint64_t result = buf[i++] - '0';
  if (unlikely(result > 9) { return NUMBER_ERROR; }
  if (result != 0) {
    while (buf[i] >= '0' && buf[i] <= '9') {
      result = result * 10 + (buf[i++] - '0');
    }
    if (unlikely((i - *element_structural_index) >= 19) {
      return parser.implementation->parse_large_integer(buf);
    }
  }
  if (unlikely(!is_whitespace_or_structural(buf[i+1]))) { return NUMBER_ERROR; }
  return result;
}
```

You can imagine how this can be specialized and vectorized for smaller things like 1 digit or 1-3 digit numbers.

### Example: Array of Structs

One can also imagine code for structs, especially if we provide methods like object_field_iterator(["field1","field2","field3"]) which gave you the fields you asked for, in order (even if they aren't in order in the JSON):

```c++
struct person_t {
  string name;
  uint64_t age;
  double weight;
}

for (auto person : simdjson::stream::parse(json)) {
  auto fields = person.get_object_fields("name","age","weight");
  person_t person_struct;
  person_struct.name = fields[0].get<string>();
  person_struct.age = fields[1].get<uint64_t>();
  person_struct.weight = fields[2].get<double>();
  process(person_struct);
}
```

Because the above code puts the fields in order, the inlined parsing code for all three would be one after the other with no branches between (besides an error branch). One can imagine branchless ways of writing get_object_fields.

Open Questions
--------

### Code Size

Exactly how many JSON fields are there in a typical ? Will the inlining of the parsing code get to be a bit too much? Would functions for parsing each type work OK, or a combination of "critical inlining that is likely to get optimized" and functions?

### Validation

There is a question of validation here: parsing and validating go hand in hand, so if we want to efficiently avoid parsing everything, we have to avoid *validating* some of it too. Full validation should likely be a default, because it's a part of the JSON specification (and for good reasons).

I don't have a full answer here. It probably involves a combination of increased prevalidation, and lazy validation of values the user doesn't look at. Obviously, both of these can decrease performance, but it depends on how and where you do them. One bright spot: validation is a lot easier than parsing in many cases, and probably even vectorizable. Think numbers: you only have to validate the form, since "number doesn't fit in 64 bits" is still valid JSON).

A "relaxed validation mode" becomes possible as well, though. We validate only the JSON necessary to read the values you ask for, but fullly validate that: in other words, we guarantee that the data you want is valid and means what the user says it means. It probably means we need to validate all the open and close brackets on the way to the value, but it seems likely we'll want to do that in any case.

### SIMD and inlining

One problem here is that we're inlining parsing code to the user, and some of that is implementation-specific. In particular, the string parsing code in stage 2, and a little bit of the number parsing code, uses simd operations. We don't have a way to inline implementation-specific code into user code, because it'd require them to have a version of their own code for each implementation.

Ultimately, we'll need the user to be able to hook into the implementation mechanism when they really want that last bit of performance. This is a solvable problem but outside our current scope. We'll probably start by providing least-common-denominator methods, but calling out to virtual .
Solving this in a way that takes unique advantage of all architectures is beyond the scope of this feature. We'll provide a least-common-denominator version (maybe SSE2 or ARMv7 depending on compiler) and (for extra credit) call an implementation-specific method for massive strings only. This may slow things down; we'll just have to see if it slows them down enough. There is reason to hope it won't: strings don't generally get that big, and numbers *can't* get bigger than 19 digits so 16-byte SIMD is more than enough for that.

Future
-------

It does not solve the problem of fast element counts needed for allocation. Depending on how this works out, we may want to go deeper and parse out objects and arrays in stage 1 (this could help vectorization too), but it's worth seeing how far we can take forward iteration.

As a stretch, methods can be provided for things like get_ascii_string(), get<uint8_t>(). and even validation types like get_number<INT_TYPE, MIN, MAX>() that can optimize things like single digit numbers.

I think we'll end up wanting to do something special with keys. Key lookup is so common, and keys are so shared, that there has to be something we can do there.

**Comment by jkeiser** [comment]
This seems like a pretty reasonable path to a streaming API to me, as well.

**Comment by lemire** [comment]
Moved to 0.6.

**Comment by jkeiser** [comment]
Resolved through #947.
